pkg/apis/core/validation/shoot.go

```diff
@@ -938,14 +977,16 @@ func validateDNS(dns *core.DNS, fldPath *field.Path) field.ErrorList {
 	for i, provider := range dns.Providers {
 		idxPath := fldPath.Child("providers").Index(i)
 
-		if provider.SecretName != nil && provider.Type != nil {
-			providerName := gardenerutils.GenerateDNSProviderName(*provider.SecretName, *provider.Type)
+		allErrs = append(allErrs, validateDNSCredentialsRef(provider, idxPath)...)
+
+		if provider.CredentialsRef != nil && provider.Type != nil {
+			providerName := gardenerutils.GenerateDNSProviderName(provider.CredentialsRef, *provider.Type)
```

Comment:
Can we use a similar approach to https://github.com/gardener/gardener/pull/13552/files#r2588165756?

I would even delete the `GenerateDNSProviderName` function if that is not needed anywhere else as its name is confusing IMO.